### PR TITLE
fixing contiguous on broadcasted dimension

### LINKED ISSUE #1514: Expand indexing error causes incorrect results
### 🐛 Expand indexing error causes incorrect results

It seems we are indexing into the expanded tensor incorrectly. The repro works correctly when the expand is dropped and we treat the size-1 axis as a broadcast dimension.

```python
import torch
inps = [(torch.Size([2, 128]), torch.float32), (torch.Size([2, 1]), torch.float32), (torch.Size([2, 128]), torch.float32)]
inps = [torch.empty(())] + [torch.randn(shape, dtype=dtype, device='cuda') for (shape, dtype) in inps]

def forward(self, sub, div_1, mul_8):
    expand = torch.ops.aten.expand(div_1, [2, 128]);  div_1 = None
    mul_10 = torch.ops.aten.mul(expand, sub);  expand = sub = None
    add = torch.ops.aten.add(mul_10, mul_8);  mul_10 = mul_8 = None
    return (add,)

ref = forward(*inps)
f = torch.jit.script(forward)
with torch.jit.fuser("fuser2"):
  for _ in range(5):
    res = f(*inps)

for a, b in zip(ref, res):
  print((a - b).abs().max())
```

### Versions

```
PyTorch version: 1.11.0a0+git7b068db
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.18.2
Libc version: glibc-2.31

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-70-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration:
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB

Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.812
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.2
[pip3] numpydoc==1.1.0
[pip3] nvidia-dlprof-pytorch-nvtx==1.1.0
[pip3] torch==1.12.0a0+git5f7ab32
[pip3] torchaudio==0.8.1
[pip3] torchvision==0.9.1
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1514

Issue arises when we try to collapse a dimension to a broadcasted dimension. PyTorch mark all dimension with stride 1 to be contiguous. This used to be fine until recently we changed it that broadcasted dimension with stride 0 can now be put as a faster dimension. So now we can be collapsing a normal dimension into a broadcasted dimension and our index is wrong, since broadcasted dimension has stride 0.

The solution is to explicitly check inner dimension stride for contiguous dimension and only collapse when inner dimension is not broadcasted.

### CODE REVIEW

**Comment by jjsjann123** [comment]
I'm awaiting local tests to finish BTW... I'll merge this once that passed and approved as well.

**Review by naoyam — APPROVED** [review]
Thanks for the fix.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 1]
```diff
@@ -96,12 +96,19 @@ TensorView::TensorView(
...
-                    ->stride_index_) {
-          // collapse if two axes are neighboring in both sizes & stride_index;
-          contig_info[index] = (left_index_opt.value() == (index + 1));
+        // check the neighboring faster dimension, collapse if two axes are
+        // neighboring in stride_index;
+        auto left_index_opt =
+            tensor_type->stride_properties()[static_cast<int>(i) - 1]
+                ->stride_index_;
+        if (left_index_opt.has_value() &&
+            left_index_opt.value() == (index + 1)) {
```
Comment (ngimel) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: how come you are collapsing based only on order of the strides, and not strides and sizes? What if tensor is not memory dense?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Totally not expecting to run into you here :)

But we did peek contiguous flag, which is supposed to check stride/size to ensure neighboring dimensions are memory dense: https://github.com/csarofeen/pytorch/pull/1519/files/0372548daf10ace4da6dddf1c1fc5c6ca8ce4b5c#diff-28d672948f070384f5d8c15ec70180f5ac1b7632d713af96dec1634fbdbb89ceR91

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Not sure if the link is working for you but here we go vvv
```
  for (const auto i : c10::irange(tensor_type->dim().value())) {
    // if we don't have contiguous dimension at current stride index, don't
    // bother;
    const auto& stride_property_i = tensor_type->stride_properties()[i];
    if (stride_property_i.has_value() &&
        stride_property_i->stride_index_.has_value() &&
        stride_property_i->contiguous_.has_value() &&
        stride_property_i->contiguous_.value() == true) {
      const size_t index = stride_property_i->stride_index_.value();
```

**Comment by jjsjann123** [comment]
FYI, tests are passing, I'm trying to do a quick write up explaining stride_properties() before merging this

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 1 above)
Comment (ngimel) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Ok, then my question is should `contiguous_` value be fixed in `computeStrideProps`? There's
```
        contiguous_ = strides[stride_indices[i]] == 1 ||
            (strides[stride_indices[i]] != 0 &&
             strides[stride_indices[i]] ==
                 strides[stride_indices[i - 1]] * sizes[stride_indices[i - 1]]);
 ```
 why is stride of 1 enough to mark something contiguous? I'd understand if size of 1 was.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: > why is stride of 1 enough to mark something contiguous? I'd understand if size of 1 was.

This is a tricky topic. I think the assumption is that we won't have overlapped memory, so `stride == 1` means that current dimension would be the fastest dimension with non-trivial length, while implicitly broadcasted dimensions are technically faster but not collapsible.

In reality I don't think we are efficiently leveraging this yet....

Comment (ngimel) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Why do you have an assumption that you won't have overlapped memory?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: That's a fair ask. we (nvfuser) don't have determined outputs for overlapped tensor, because of the racing conditions on write. So we don't care much about it. :stuck_out_tongue: 

In case of contiguity map for overlapped memory, it's also hard/tricky to figure out which dimensions are overlapping in the general case. reference:  https://github.com/pytorch/pytorch/blob/0be36d798ba959bfda6c448fc4832b5691df6e61/aten/src/ATen/MemoryOverlap.h#L12-L19
I do agree that we can be a little bit more cautious/conservative on this and mark all contiguity as false for YES/TOO_HARD cases.

Comment (ngimel) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: I understand, but in this case contiguity map was wrong for an input tensor, not output tensor, and input tensors are allowed to be non-overlapping. General overlapping is hard, but contiguity isn't.
